iris-mpc-gpu/tests/e2e.rs

```diff
@@ -360,15 +363,54 @@ mod e2e_test {
                                 right: db.db[deleted_idx as usize].clone(),
                             }
                         }
-                        // TODO: implement test that enables testing the bitmap OR rule setting.
-                        // This will require setting the left and right iris code to be different,
-                        // one of them below the threshold and the other
-                        // above the threshold against the other codes. It will
-                        // also be required to set the param BatchQuery.threshold_bitmap_or to true
-                        // for the specific iris codes to be matched against
-                        // 5 => {
-                        // println!("Sending iris codes that match on right but not left with the OR
-                        // rule set");
+                        5 => {
+                            println!(
+                                "Sending iris codes that match on right but not left with the OR \
+                                 rule set"
+                            );
+                            let db_index = loop {
+                                let db_index = rng.gen_range(0..DB_SIZE / 10);
+                                if !disallowed_queries.contains(&db_index) {
+                                    break db_index;
+                                }
+                            };
+                            if deleted_indices.contains(&(db_index as u32)) {
+                                continue;
+                            }
+
+                            // comparison against this item will use the OR rule
+                            use_or_rule_for_serial_ids.push(db_index as u32);
+
+                            // Will always match under the OR rule
+                            expected_results
+                                .insert(request_id.to_string(), (Some(db_index as u32), false));
+
+                            let mut code_left = db.db[db_index].clone();
+                            let mut code_right = db.db[db_index].clone();
+
+                            assert_eq!(code_left.mask, IrisCodeArray::ONES);
+                            assert_eq!(code_right.mask, IrisCodeArray::ONES);
+
+                            // apply variation to either right of left code
+                            let flip_right: bool = rng.gen();
+                            let variation = 1;
+
+                            if flip_right {
+                                // Flip bits to below threshold
+                                for i in 0..(THRESHOLD_ABSOLUTE as i32 - variation) as usize {
```

Comment:
I think this actually flips to few bits, since the variation is positive so we only flip less than the Threshold. This means that these would also match using the AND rule if I am not mistaken.

Reply:
Maybe a better test case would be to add the query to the batch twice, once with AND rule, once with OR rule and check the two results are as expected. However, IDK what the handling of the OR rule with the inter-batch checks is yet, so IDK if this would even work out.

Reply:
See above: https://github.com/worldcoin/iris-mpc/pull/951/files/424926dbd45f45c6bb1ae96a948cdab185c42aad#diff-0c1888f81c1fbf6d412eaebb8144b3f97d10ae89d279ca4153b4f900003a69bdR321-R330

When variation is < 0 (-1) we run:

```
for i in 0..(THRESHOLD_ABSOLUTE as i32 + variation) as usize {
                                code.code.flip_bit(i);
                            }
```

Which is the same as I have here, since I set variation fixed to `let variation = 1`

I did verify that the results do indeed come as `true` when flipping the variation value when running the tests, fwiw.

But I will think a bit more deeply on how to implement a more comprehensive test

Reply:
but that is my point, no? If variation is -1, then in the branch for 2, we go into the else path which is an expected match, since we flip less than the threshold. I thought what you want here is to flip enough bits in one half so that it no longer is a match?

Reply:
Though at the threshold it should match anyways right? We use =< IIRC, so flipping one side to less than the threshold will leave the other matching  and the one with the flip bits non-matching. 

It could be that my reasoning was mistaken, but the results were correct? I'll refactor to flip one side to threshold + 1 and other side threshold - 1 for clarity

Reply:
An iris code is a match if the fractional distance is less-or-equal than the threshold 0.375 or 3/8. 
When copying an iris code from the DB, this means that the default distance between this query and the position in the DB is 0, so it is a match. To make it a non-match, we need to flip more than 3/8 of the bits to cause it to not match. This threshold_absoute value is exactly 3/8*IRIS_CODE_LEN, so flipping bits up to and including this value will leave the distance less-or-equal to 3/8, which still is a match. Only flipping more bits than this value will result in a non-match. (This also assumes that all of the iris code masks are set to 1, which I think is the case here.